### ISSUE TITLE
Solve Time Variable in the Substitution Action

### ISSUE THREAD

**Issue by rgrinberg** [ocaml/dune#9402]
Variables such as `%{with-test}` are determined at solve time. These variables should be correctly substituted at build time. This is done via partial evaluation in the lock files, but we end up forgetting the values for these variables when running the substitution action.

Now that @gridbugs started recording the values for all these variables in the lock directory, we can start properly using them in substitute action as well.

**Comment by rgrinberg** [comment]
@gridbugs feel free to assign this to yourself if you plan doing this

**Comment by gridbugs** [comment]
Why not substitute these variables at solve-time instead? We already substitute the variables in the filters on build and install commands and then apply those commands, so the why not extend this to the variables referred to in the commands themselves?

**Comment by rgrinberg** [comment]
B/c it's not possible. The variables we need to substitute in the `(substitute ..)` action are done on the package sources. So we don't know what to substitute until the build starts.

**Comment by gridbugs** [comment]
Makes sense. The only variables that get recorded in the lockdir are those used while solving. I imagine we want to use the environment on the current system at build time to substitute these variables, and use the variables stored in the lockdir to validate that the current system is compatible with the lockdir.

**Comment by rgrinberg** [comment]
> and use the variables stored in the lockdir to validate that the current system is compatible with the lockdir.

That would address #9406. Although I'm still on the fence whether it's a good idea.

**Comment by gridbugs** [comment]
Taking a closer look at this I'm hesitant to commit to a solution until we have a clear story about https://github.com/ocaml/dune/issues/9406. As it stands, the solver variables that we poll from the system are already pform variables. If variables recorded from the solver env differ from the current system then I presume we just want to take their current values rather than their values at solve time (assuming that we decide to attempt to build with lockfiles that differ from the current system at all).

Something else I'm thinking about is what to do about solver variables that don't have corresponding pform variables. Users are free to define whatever solver variables they like in the workspace currently I think these are just used during solving. We can completely substitute occurrences of these variables in filters and commands at solve-time, but I suppose we'd also want to substitute these in the `substitute` action too?
